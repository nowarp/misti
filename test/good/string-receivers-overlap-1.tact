const WHATEVER: Bool = true;

const a: String = "foo";

contract Test {
    receive("test") {}
    receive(msg: String) {
        // Simple cases
        if (msg == "test") {} // Bad
        if ("test" != msg) {} // Bad
        if (("test" != msg) && WHATEVER) {} // Bad
        if ((WHATEVER || WHATEVER) && ("test" != msg)) {} // Bad

        // Local variables tainted with `msg`
        let foo: String = msg;
        let bar = foo;
        if ("test" == bar) {} // Bad

        // Local variables tainted with `"test"`
        let test1: String = "test";
        let test2 = test1;
        if (test2 == msg) {} // Bad

        // Combining tainted cases
        if (test2 != bar) {} // Bad
        if (bar == test2) {} // Bad

        // No false positives
        if ("test" != a) {}
    }
}
