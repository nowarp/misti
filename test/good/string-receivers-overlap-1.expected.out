test/good/string-receivers-overlap-1.tact:9:13:
   8 |         // Simple cases
>  9 |         if (msg == "test") {} // Bad
                   ^
  10 |         if ("test" != msg) {} // Bad
String Receivers Overlap: receiver("msg") might be called instead. This condition might never be executed.
Help: Implement the desired logic in receiver("msg") and remove msg == "test"
See: https://nowarp.github.io/tools/misti/docs/detectors/StringReceiversOverlap

test/good/string-receivers-overlap-1.tact:10:13:
   9 |         if (msg == "test") {} // Bad
> 10 |         if ("test" != msg) {} // Bad
                   ^
  11 |         if (("test" != msg) && WHATEVER) {} // Bad
String Receivers Overlap: receiver("msg") might be called instead. This condition might never be executed.
Help: Implement the desired logic in receiver("msg") and remove "test" != msg
See: https://nowarp.github.io/tools/misti/docs/detectors/StringReceiversOverlap

test/good/string-receivers-overlap-1.tact:11:13:
  10 |         if ("test" != msg) {} // Bad
> 11 |         if (("test" != msg) && WHATEVER) {} // Bad
                   ^
  12 |         if ((WHATEVER || WHATEVER) && ("test" != msg)) {} // Bad
String Receivers Overlap: receiver("msg") might be called instead. This condition might never be executed.
Help: Implement the desired logic in receiver("msg") and remove "test" != msg
See: https://nowarp.github.io/tools/misti/docs/detectors/StringReceiversOverlap

test/good/string-receivers-overlap-1.tact:12:13:
  11 |         if (("test" != msg) && WHATEVER) {} // Bad
> 12 |         if ((WHATEVER || WHATEVER) && ("test" != msg)) {} // Bad
                   ^
  13 | 
String Receivers Overlap: receiver("msg") might be called instead. This condition might never be executed.
Help: Implement the desired logic in receiver("msg") and remove "test" != msg
See: https://nowarp.github.io/tools/misti/docs/detectors/StringReceiversOverlap

test/good/string-receivers-overlap-1.tact:17:13:
  16 |         let bar = foo;
> 17 |         if ("test" == bar) {} // Bad
                   ^
  18 | 
String Receivers Overlap: receiver("msg") might be called instead. This condition might never be executed.
Help: Implement the desired logic in receiver("msg") and remove "test" == bar
See: https://nowarp.github.io/tools/misti/docs/detectors/StringReceiversOverlap

test/good/string-receivers-overlap-1.tact:22:13:
  21 |         let test2 = test1;
> 22 |         if (test2 == msg) {} // Bad
                   ^
  23 | 
String Receivers Overlap: receiver("msg") might be called instead. This condition might never be executed.
Help: Implement the desired logic in receiver("msg") and remove test2 == msg
See: https://nowarp.github.io/tools/misti/docs/detectors/StringReceiversOverlap

test/good/string-receivers-overlap-1.tact:25:13:
  24 |         // Combining tainted cases
> 25 |         if (test2 != bar) {} // Bad
                   ^
  26 |         if (bar == test2) {} // Bad
String Receivers Overlap: receiver("msg") might be called instead. This condition might never be executed.
Help: Implement the desired logic in receiver("msg") and remove test2 != bar
See: https://nowarp.github.io/tools/misti/docs/detectors/StringReceiversOverlap

test/good/string-receivers-overlap-1.tact:26:13:
  25 |         if (test2 != bar) {} // Bad
> 26 |         if (bar == test2) {} // Bad
                   ^
  27 | 
String Receivers Overlap: receiver("msg") might be called instead. This condition might never be executed.
Help: Implement the desired logic in receiver("msg") and remove bar == test2
See: https://nowarp.github.io/tools/misti/docs/detectors/StringReceiversOverlap

test/good/string-receivers-overlap-1.tact:17:23:
  16 |         let bar = foo;
> 17 |         if ("test" == bar) {} // Bad
                             ^
  18 | 
Read-only variable
Help: Consider creating a constant instead
See: https://nowarp.github.io/tools/misti/docs/detectors/ReadOnlyVariables

test/good/string-receivers-overlap-1.tact:22:13:
  21 |         let test2 = test1;
> 22 |         if (test2 == msg) {} // Bad
                   ^
  23 | 
Read-only variable
Help: Consider creating a constant instead
See: https://nowarp.github.io/tools/misti/docs/detectors/ReadOnlyVariables